internal/component/otelcol/exporter/datadog/datadog.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/grafana/alloy/internal/component/otelcol/exporter"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter"
+	otelcomponent "go.opentelemetry.io/collector/component"
+	otelextension "go.opentelemetry.io/collector/extension"
+)
+
+const (
+	DATADOG_TRACE_ENDPOINT   = "https://trace.agent.%s"
+	DATADOG_METRICS_ENDPOINT = "https://api.%s"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:      "otelcol.exporter.datadog",
+		Community: true,
+		Args:      Arguments{},
+		Exports:   otelcol.ConsumerExports{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			fact := datadogexporter.NewFactory()
+			return exporter.New(opts, fact, args.(Arguments), exporter.TypeAll)
+		},
+	})
+}
+
+// Arguments configures the otelcol.exporter.datadog component.
+type Arguments struct {
+	Client  HTTPClientArguments    `alloy:"client,block,optional"`
+	Timeout time.Duration          `alloy:"timeout,attr,optional"`
+	Queue   otelcol.QueueArguments `alloy:"sending_queue,block,optional"`
+	Retry   otelcol.RetryArguments `alloy:"retry_on_failure,block,optional"`
+
+	// Datadog specific configuration settings
+	APISettings  otelcol.DatadogAPIArguments          `alloy:"api,block"`
+	Traces       otelcol.DatadogTracesArguments       `alloy:"traces,block,optional"`
+	Metrics      otelcol.DatadogMetricsArguments      `alloy:"metrics,block,optional"`
+	HostMetadata otelcol.DatadogHostMetadataArguments `alloy:"host_metadata,block,optional"`
+	OnlyMetadata bool                                 `alloy:"only_metadata,attr,optional"`
+	Hostname     string                               `alloy:"hostname,attr,optional"`
+
+	// DebugMetrics configures component internal metrics. Optional.
+	DebugMetrics otelcolCfg.DebugMetricsArguments `alloy:"debug_metrics,block,optional"`
+}
+
+// DatadogAPISettings holds the configuration settings for the Datadog API.
+
+var _ exporter.Arguments = Arguments{}
+
+// SetToDefault implements syntax.Defaulter.
+func (args *Arguments) SetToDefault() {
+	args.APISettings.SetToDefault()
+	args.Metrics.SetToDefault()
+	args.Traces.SetToDefault()
+	args.HostMetadata.SetToDefault()
+
+	args.Queue.SetToDefault()
+	args.Retry.SetToDefault()
+	args.Client.SetToDefault()
+	args.DebugMetrics.SetToDefault()
+}
+
+// Convert implements exporter.Arguments.
+func (args Arguments) Convert() (otelcomponent.Config, error) {
+	// Bubble down the API site to the metrics/traces settings. We do it this way
+	// as if we set a default, we can't tell if they specifically overrided this field as
+	// SetToDefault is called before Convert, and we need to set other defaults.
+	if args.Traces.Endpoint == "" {
+		args.Traces.Endpoint = fmt.Sprintf(DATADOG_TRACE_ENDPOINT, args.APISettings.Site)
+	}
+	if args.Metrics.Endpoint == "" {
+		args.Metrics.Endpoint = fmt.Sprintf(DATADOG_METRICS_ENDPOINT, args.APISettings.Site)
+	}
```

Comment:
I don't think that the Convert function should modify the values of the args. I would suggest to only set the values to the otel config struct. To achieve this without modifying the args you could create two variables and pass them to the convert functions of Traces.Convert() and Metrics.Convert()

Reply:
Done, thanks for the idea on that.